### ISSUE TITLE
Silent requests inadvertently serialized in MSAL 2.0.10-2.0.12, Common 3.2.0-3.4.3

### ISSUE THREAD

**Issue by AdamBJohnsonx** [AzureAD/microsoft-authentication-library-common-for-android#1438]
In MSAL version 2.0.10 and MSAL-Android-common version 3.2.0, a change was introduced that resulted in the serialization of all silent requests.  This resulted from the change in https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1230/files being adopted in commit https://github.com/AzureAD/microsoft-authentication-library-common-for-android/commit/cdb1dee78ec44c85909e23ce31a6e20a38f5f8f9.

This will be fixed in an upcoming release, tentatively scheduled for next week.  In the intervening time, **do not use** these versions of the libraries, and move to version 2.0.8 if at all possible.

**The Details**

We made a change to add naming to the threads running in various thread pools in the system so as to better be able to understand concurrency needs.  As a part of this change, we attempted to defer thread creation, and this led to the issue here.

In the ThreadUtils class, there were essentially three methods to create a pool, one which had a synchronous queue (queue size of zero), one with a bounded queue(size n), and one with an unbounded queue (size < 1).  With the added variables of core pool size and max pool size, we created the opportunity for the situation that if the *core pool size* was less than the *max pool size*, and the queue was non-zero, no more than *core pool size* threads would ever be created. 

If we look at the java source for this code: https://cs.android.com/android/platform/superproject/+/master:libcore/ojluni/src/main/java/java/util/concurrent/ThreadPoolExecutor.java;drc=d2b4b425a84c78dfe91a81774f2805d3ee278993;l=89 we find the following section, which succinctly describes the scenario (emphasis added).

```
 * When a new task is submitted in method {@link #execute(Runnable)},
 * and fewer than corePoolSize threads are running, a new thread is
 * created to handle the request, even if other worker threads are
 * idle.  If there are more than corePoolSize but less than
 * maximumPoolSize threads running, **a new thread will be created only
 * if the queue is full**. 
```

We changed the silent pool core size to 1, from 5.  Initially, this was set to 5 and 5 by dint of Executors.newFixedThreadPool(5).  When we set it to 1, with an unbounded queue, the queue was simply never full.  This meant that no new worker threads were ever created, and therefore that we were in essence running a single-threaded system.

We are fixing this by reverting the change alter the pool to name the threads.

**Comment by shahzaibj** [comment]
This seems to primarily refer to MSAL...should the issue live in the MSAL repo? https://github.com/AzureAD/microsoft-authentication-library-for-android (I do know the actual bug is in common code)

**Comment by AdamBJohnsonx** [comment]
> This seems to primarily refer to MSAL...should the issue live in the MSAL repo? https://github.com/AzureAD/microsoft-authentication-library-for-android (I do know the actual bug is in common code)

We opened this one here.  We'll be updating the MSAL readme with a link to it.

**Comment by RoarGronmo** [comment]
Simply: Has the security level, when using this versions of the library, ever been comprimized, because of this perticular problem ?

**Comment by RoarGronmo** [comment]
Secondly: Is the problem solved in MSAL 2.1.0-RC2 ?

**Comment by AdamBJohnsonx** [comment]
@RoarGronmo This is not related to any security aspects, just that we attempt to run up to 5 simultaneous silent requests with a thread pool of maximum size 5, and this was limiting the size to 1.

**Comment by AdamBJohnsonx** [comment]
This issue is fixed in the latest version of MSAL, v2.1.0.
